### ISSUE TITLE
Couple observations after installing this

### ISSUE THREAD

**Issue by mattie47** [gordonturner/stupid-remote#6]
Hi there,

Firstly, thank you for making this app. It's really cool!

I'm using it to turn on my TV when I want to airplay music to Kodi through my stereo (which due to how my setup is done is connected to the TV's audio out). 

You might also be interested in http://kodi.wiki/view/Add-on:Kodi_Callbacks

I'm using it to look at syslog events, and do an action on matching when I start playing something via airplay to kodi. With the action field, I'm calling http://osmc:8080/cec/tvOn which so far is working well!

Just wanted to point out a couple things I had to do to get this working.

Firstly, it's missing "sleep" from "node_modules". I'll get to instructions for anyone reading later

To get this working on my raspberry pi, I needed to:

**Install required apps**
```
sudo apt-get install node npm nodejs
sudo npm install -g --unsafe-perm  node-gyp@latest
sudo npm install -g --unsafe-perm  npm@latest
```

I then attempted to run as per instructions, however it would fail due to missing node_module "sleep":

```
osmc@osmc:~/stupid-remote-3.0$ DEBUG=* node ./bin/www


 express:router use /cec router +0ms
  express:router:layer new /cec +1ms
module.js:487
    throw err;
    ^

Error: Cannot find module 'sleep'
    at Function.Module._resolveFilename (module.js:485:15)
    at Function.Module._load (module.js:437:25)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/osmc/stupid-remote-3.0/routes/macro.js:7:13)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
osmc@osmc:~/stupid-remote-3.0$ npm install sleep

> sleep@5.1.0 install /home/osmc/stupid-remote-3.0/node_modules/sleep
> node-gyp rebuild

gyp ERR! build error
gyp ERR! stack Error: not found: make
gyp ERR! stack     at getNotFoundError (/usr/local/lib/node_modules/npm/node_modules/which/which.js:13:12)
gyp ERR! stack     at F (/usr/local/lib/node_modules/npm/node_modules/which/which.js:68:19)
gyp ERR! stack     at E (/usr/local/lib/node_modules/npm/node_modules/which/which.js:80:29)
gyp ERR! stack     at /usr/local/lib/node_modules/npm/node_modules/which/which.js:89:16
gyp ERR! stack     at /usr/local/lib/node_modules/npm/node_modules/which/node_modules/isexe/index.js:42:5
gyp ERR! stack     at /usr/local/lib/node_modules/npm/node_modules/which/node_modules/isexe/mode.js:8:5
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:152:21)
gyp ERR! System Linux 4.9.29-8-osmc
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/osmc/stupid-remote-3.0/node_modules/sleep
gyp ERR! node -v v8.2.1
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! sleep@5.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sleep@5.1.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/osmc/.npm/_logs/2017-07-28T06_20_07_433Z-debug.log

```

As I've only first used node etc for the first time a week ago, I went digging for what I needed to do to get things working.

The short story is I first tried to install sleep globally to node using npm. However this I found was failing due to other missing requirements. Specifically, "make" and "g++".

So to solve this I:

```
sudo apt-get install make g++
````

Attempting to install sleep again worked:

```
sudo npm install --unsafe-perm --verbose -g sleep
```

However, for some reason stupid-remote didn't seem to just pick this up from the global node library, and I did the following to get it to work:

```
osmc@osmc:~/stupid-remote-3.0$ cd node_modules/
osmc@osmc:~/stupid-remote-3.0/node_modules$ npm install sleep

> sleep@5.1.0 install /home/osmc/stupid-remote-3.0/node_modules/sleep
> node-gyp rebuild

make: Entering directory '/home/osmc/stupid-remote-3.0/node_modules/sleep/build'
  CXX(target) Release/obj.target/node_sleep/sleep.o
  SOLINK_MODULE(target) Release/obj.target/node_sleep.node
  COPY Release/node_sleep.node
make: Leaving directory '/home/osmc/stupid-remote-3.0/node_modules/sleep/build'
npm notice created a lockfile as package-lock.json. You should commit this file.
+ sleep@5.1.0
added 2 packages in 24.823s
osmc@osmc:~/stupid-remote-3.0/node_modules$
```

Finally at this point I was able to get it working :-) 

I don't understand node enough to do a Pull Request and submit the "sleep" module, but perhaps you could add it yourself so it works out of the box for anyone else trying this.

Thanks,

Matt

**Comment by gordonturnerolson** [comment]
Thanks for the feedback Matt!

I will not have a chance to reproduce your issue and fix this weekend, hopefully next week though.

Can you tell me what version of Linux you are running?  Raspbian?  and perhaps what release?  Also what hardware?

It would help me to reproduce and update my instructions.

Gord.

**Comment by gordonturner** [comment]
(Posted with wrong Github account, doh.)

Thanks for the feedback Matt!

I will not have a chance to reproduce your issue and fix this weekend, hopefully next week though.

Can you tell me what version of Linux you are running? Raspbian? and perhaps what release? Also what hardware?

It would help me to reproduce and update my instructions.

Gord.

**Comment by mattie47** [comment]
Hi @gordonturner 

Thanks for the quick reponse!

Using OSMC which is an image for using Kodi on a Raspberry Pi3.

[2017.06-1](http://download.osmc.tv/installers/diskimages/OSMC_TGT_rbp2_20170705.img.gz) - https://osmc.tv/download/
```
osmc@osmc:~$ uname -a
Linux osmc 4.9.29-8-osmc #1 SMP PREEMPT Fri Jun 16 21:37:12 UTC 2017 armv7l GNU/Linux
```


I'm having another weird issue which I need to try look into more, where as soon as I start stupid-remote, I can no longer use my TV's Remote to control my Pi3.

Killing the stupid-remote node PID doesn't fix the issue, and it results in me restarting my Pi. 

I don't know if it's a conflict between Kodi and stupid-remote trying to use HDMI-CEC at the same time or what, but if I get some time I'll try and have a look into it more.

You haven't possibly had any weird issues like this? (From memory you've got kodi as one of the default remote options). 

Thanks,

Matt

**Comment by gordonturner** [comment]
The good news is that have a Pi running OSMC with Stupid Remote on it (using it with a usb-cec device and flirc ir adapter), so it can work.

I did seem some similar behaviour to what you are describing, my resolution at the time was to try and kill things and then restart.

I *think* what fixed it was to get a more current version of cec-client.  IIRC the version that ships with Raspbian / OSMC was not current.  

I followed these instructions:
https://github.com/Pulse-Eight/libcec/blob/master/docs/README.raspberrypi.md

Obviously something is not letting go of the cec device, so when something else (Kodi) wants to use it, things fail.

Try that out?
Gord.

**Comment by mattie47** [comment]
Ah - Thanks for that Gordon. 

I'll try that some stage later this week!

Will report back with how I go...

From the topic of home automation - You might find this interesting: [Broadlink RM3 Mini](https://www.aliexpress.com/item/2016-New-Original-Broadlink-RM-Mini3-Universal-Intelligent-WiFi-IR-4G-Wireless-Remote-Controller-Via-Phone/32665685151.html?spm=2114.10010108.1000014.5.1037eac5IKY0lD&traffic_analysisId=recommend_3035_null_null_null&scm=1007.13338.80878.000000000000000&pvid=879635a7-1f14-4ad6-8031-5db1ef47d4d8&tpp=1)

I've been using these to turn heaters off/on remotely but of course you could also use it to turn the different devices on your TV off/on. 
 
Thanks,

Matt

**Comment by mattie47** [comment]
Hi @gordonturner 

Firstly thanks for the help.

I compiled libcec tonight mostly following the instructions on that link (ran into a couple issues which I documented here: https://github.com/Pulse-Eight/libcec/issues/373)

Anyway, sadly that doesn't appear to have fixed my issue.

I should point out after trying above, I don't think the issue is in your app. I came across another app for the same-ish task https://github.com/dan-nguyen/hdmi-cec-node-test which I also get the same issue.

I need to try turning on kodi debug again, and try get an idea of what isn't happy.

If I find anything useful, I'll report back.

Thanks,

Matt

**Comment by gordonturner** [comment]
Hey Matt,

I have updated the notes for installing stupid-remote (on the develop branch):

https://github.com/gordonturner/stupid-remote/blob/develop/README%20Install%20Raspbian.md

https://github.com/gordonturner/stupid-remote/blob/develop/README%20Install%20OSMC.md

Have a look at the OSMC one and please let me know if those instructions help you out?

Thanks!
Gord.

**Comment by mattie47** [comment]
Hey Gordon,

I've given both docs a quick read and they look good. 

Reading in between the lines, was there something I failed to do to get node_module "sleep"? 

I'm just wondering where that issue has been addressed? 

And I haven't given up on this :-) I'm just stumped as to how to achieve my goal with Kodi, as the issue from what I can see is that it takes ownership of CEC functionality using libcec. As soon as Stupid-remote activates and does something, I see in the kodi debug *something* terminate which I can only assume is the CEC functionality.

Essentially I guess my real issue here is that kodi-core does not expose the CEC functionality to their API (which frankly on their part would be quite useful).

Thanks,

Matt

**Comment by gordonturner** [comment]
Hey Matt,

The issue (as I understand it) that you were seeing I was able to reproduce.

It was caused because OSMC / Raspbian ship with an old version of nodejs.  The work around (as per the updated instructions) is to remove the old version and install the version from the node js project.

Are you using the TV remote to control OSMC?

I am not using my TV remote to control the Pi, so I have it disabled in OSMC (just confirmed it).  If that is an option for you that should work.

Otherwise, if you are using the TV remote, I haven't really explored how to share the cec device between different apps (OSMC and stupid-remote).  

G.

**Comment by mattie47** [comment]
Hey Gordon,

> It was caused because OSMC / Raspbian ship with an old version of nodejs. The work around (as per the updated instructions) is to remove the old version and install the version from the node js project.

Ah right. This makes sense. 

> I am not using my TV remote to control the Pi, so I have it disabled in OSMC (just confirmed it). If that is an option for you that should work.

Riiiight okay that makes a difference in set ups. Yes I am using my TV's remote to control Kodi.

No worries and thanks for the help. Again if I find anything else useful I'll post it here, but I think at this point I can close the comment.


Thanks,

Matt

**Comment by mattie47** [comment]
Hey @gordonturner ,


Thought it was worth updating what I ended up doing.

Kodi takes control of CEC functionality through libCEC, so calling libcec or cec-client separately causes the remote issue I previously mentioned.

Previously I said:

> Essentially I guess my real issue here is that kodi-core does not expose the CEC functionality to their API (which frankly on their part would be quite useful).

This isn't entirely true.

Kodi actually has a number of built in functions - http://kodi.wiki/view/list_of_built-in_functions

3 of which are:

CECActivateSource | Wake up playing device via a CEC peripheral |
-- | -- |
CECStandby | Put playing device on standby via a CEC peripheral | 
CECToggleState | Toggle state of playing device via a CEC peripheral |

So while these aren't exposed through a specific Kodi API as such, there is another Kodi plugin which allows you to send these through JSON-RPC:

https://github.com/joshjowen/script.json-cec

So I've now got this doing what I wanted :-)

Thanks,

Matt
